### PR TITLE
Spread call/eth

### PR DESCRIPTION
**Pull request by ugolino**

### CODE REVIEW

File: contracts/actions/ShortOTokenActionWithSwap.sol [hunk 1]
```diff
@@ -195,17 +181,17 @@ contract ShortOTokenActionWithSwap is IAction, AirswapBase, RollOverBase {
   function flashMintAndSellOToken(uint256 optionsToSell, uint256 premium, address counterparty) external onlyOwner { 
     //0. Initial Logic Checks
     //require(counterparty.add != address(0), "Invalid counterparty address");
```
Comment (aparnakr) [inline contracts/actions/ShortOTokenActionWithSwap.sol:183]: probs need to uncomment this and implement a signature scheme to make sure the counterparty is ok with this transaction

File: contracts/actions/ShortOTokenActionWithSwap.sol [hunk 2]
```diff
@@ -195,17 +181,17 @@ contract ShortOTokenActionWithSwap is IAction, AirswapBase, RollOverBase {
...
-    
+
     // flash borrow WETH
     address receiverAddress = address(this);
     address[] memory assets = new address[](1);
     assets[0] = address(weth);
     uint256[] memory amounts = new uint256[](1);
     uint256 collateralNeeded = optionsToSell.mul(1e10);
```
Comment (aparnakr) [inline contracts/actions/ShortOTokenActionWithSwap.sol:190]: not needed right now, but theoretically you could borrow collateralNeeded - premium, that way pay less fees on the flash loan.

File: contracts/actions/ShortOTokenActionWithSwap.sol [hunk 3]
```diff
@@ -195,17 +181,17 @@ contract ShortOTokenActionWithSwap is IAction, AirswapBase, RollOverBase {
...
-    uint256 amountSdEcrvInAction = stakedao.balanceOf(address(this));
+    uint256 collateralInAction = weth.balanceOf(address(this));
     
     // sdcrv needed
```
Comment (aparnakr) [inline contracts/actions/ShortOTokenActionWithSwap.sol:193]: weth needed

File: contracts/core/OpynPerpVault.sol [hunk 4]
```diff
@@ -119,15 +112,14 @@ contract OpynPerpVault is ERC20, ReentrancyGuard, Ownable {
    *====================*/
 
   constructor (
-    address _sdecrvAddress,
-    address _curvePool,
+    address _asset,
+    // address _curvePool,
```
Comment (aparnakr) [inline contracts/core/OpynPerpVault.sol:116]:
```suggestion
```

File: contracts/core/OpynPerpVault.sol [hunk 5]
```diff
@@ -170,14 +162,14 @@ contract OpynPerpVault is ERC20, ReentrancyGuard, Ownable {
...
-    return totalStakedaoAsset().mul(sdecrv.getPricePerFullShare()).mul(curvePool.get_virtual_price()).div(10**36);
-  }
+  // /**
+  //  * total eth value of the sdecrv controlled by this vault
+  //  */
+  // function totalETHControlled() external view returns (uint256) { 
+  //   IStakeDao sdecrv = IStakeDao(sdecrvAddress);
+  //   // hard coded to 36 because ecrv and sdecrv are both 18 decimals. 
+  //   return totalAsset().mul(sdecrv.getPricePerFullShare()).mul(curvePool.get_virtual_price()).div(10**36);
+  // }
```
Comment (aparnakr) [inline contracts/core/OpynPerpVault.sol:172]:
```suggestion
```

File: contracts/core/OpynPerpVault.sol [hunk 6]
```diff
@@ -234,38 +209,41 @@ contract OpynPerpVault is ERC20, ReentrancyGuard, Ownable {
...
-    uint256 fee = _getWithdrawFee(ethReceived);
-    uint256 ethOwedToUser = ethReceived.sub(fee);
+    uint256 fee = _getWithdrawFee(assetToWithdraw);
+    uint256 ethOwedToUser = assetToWithdraw.sub(fee);
+
+    IWETH(asset).withdraw(assetToWithdraw);
 
     // send fee to recipient 
     (bool success1, ) = feeRecipient.call{ value: fee }('');
     require(success1, 'O9');
```
Comment (aparnakr) [inline contracts/core/OpynPerpVault.sol:230]: could just send fee in ETH, though not a big issue

File: contracts/core/OpynPerpVault.sol [hunk 7]
```diff
@@ -414,10 +392,10 @@ contract OpynPerpVault is ERC20, ReentrancyGuard, Ownable {
...
-    */
-  receive() external payable {
-    require(msg.sender == address(curvePool), "O19");
-  }
+  // /**
+  //   * @notice the receive ether function is called whenever the call data is empty
+  //   */
+  // receive() external payable {
+  //   require(msg.sender == address(curvePool), "O19");
+  // }
```
Comment (aparnakr) [inline contracts/core/OpynPerpVault.sol:400]:
```suggestion
```

File: test/mainnet-fork/short-vault-loss.ts [hunk 8]
```diff
@@ -165,11 +154,10 @@ describe('Mainnet Fork Tests', function() {
     );
     action1 = (await ShortActionContract.deploy(
       vault.address,
-      stakeDaoTokenAddress,
+      // wethAddress,
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:157]: why commented out?

Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:157]:
```suggestion
```

File: test/mainnet-fork/short-vault-loss.ts [hunk 9]
```diff
@@ -218,58 +195,38 @@ describe('Mainnet Fork Tests', function() {
...
       .whitelistProduct(
         weth.address,
         usdc.address,
-        stakeDaoTokenAddress,
+        wethAddress,
         false
       );
     await provider.send('evm_mine', []);
     await provider.send('hardhat_stopImpersonatingAccount', [opynOwner]);
   });
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:222]: you probably dont need this because weth is already whitelisted in the system

File: test/mainnet-fork/short-vault-loss.ts [hunk 10]
```diff
@@ -365,20 +322,20 @@ describe('Mainnet Fork Tests', function() {
...
-      const totalSharesMinted = vaultSdecrvBalance;
+      const vaultTotal = await vault.totalAsset();
+      const vaultwethBalance = await weth.balanceOf(vault.address);
+      const totalSharesMinted = vaultwethBalance;
 
-      // check the sdeCRV token balances
+      // check the weth token balances
       expect(
-        (vaultTotal).gte(expectedSdecrvInVault),
+        (vaultTotal).gte(expectedwethInVault),
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:335]: since this is weth and no stakedao and curve fees, you should just be able to check that it is equal! No need to approx / estimate :)

File: test/mainnet-fork/short-vault-loss.ts [hunk 11]
```diff
@@ -389,37 +346,36 @@ describe('Mainnet Fork Tests', function() {
 
     it('p2 deposits', async () => {
       // there is no accurate way of estimating this, so just approximating for now
-      const expectedSdecrvInVault = p1DepositAmount.mul(95).div(100);
+      const expectedwethInVault = p1DepositAmount.mul(95).div(100);
```
Comment (aparnakr) [inline test/mainnet-fork/short-vault-loss.ts:349]: same thing, no need to approx, you can just check that this is equal because no intermediate fees that we aren't aware of.

**Review by aparnakr — APPROVED** [review]
Generally good direction! 

Left some comments to make changes :) 

TLDR:
1. We should implement a signature scheme to make sure the counterparty approves this particular amount of options, this spread and this premium
2. In the tests, since its all in weth and there is no hidden fees, its better to check the exact amount than to estimate / approximate values
